Fix dynamic shapes test failures on SM 7.x or lower.

alexsamardzic opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123762

For details, see: https://github.com/pytorch/pytorch/issues/122747.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

alexsamardzic committed:
Fix dynamic shapes test failures on SM 7.x or lower.

For details, see: https://github.com/pytorch/pytorch/issues/122747.

[ghstack-poisoned]

alexsamardzic commented:
Fixes #122747.

eqy commented:
Assigning myself to this to check if the failure happens on A2/A16 (GA107) which is technically sm80+ but not commonly tested.

alexsamardzic commented:
> Assigning myself to this to check if the failure happens on A2/A16 (GA107) which is technically sm80+ but not commonly tested.

There is no reason to fail on anything that is 8.x or higher, as the failure was reported exactly because of (valid) code path that is specific for 7.x and lower; but - yeah, it won't hurt to check if you have access to this GPU.  FWIW, it passes on my laptop GA104.

alexsamardzic commented:
Ok, so decorator should be changed to apply only if test run for "cuda" device.  Not sure how to do that...

eqy commented on test/inductor/test_torchinductor.py:
```diff
@@ -134,6 +134,11 @@
 
+def expectedFailureCodegenDynamicIfNotSM80OrLater(fn):
+    if not SM80OrLater:
+        fn._expected_failure_codegen_dynamic = True
```
Would it be possible to pattern match the implementation of e.g., `dtypesIfCUDA` since this sounds like "expected failure codegen dynamic if cuda" https://github.com/pytorch/pytorch/blob/e00282fecfcb53790aebfb24cc48a8703577778e/torch/testing/_internal/common_device_type.py#L1278

alexsamardzic replied:
Yep, let's try something alike.

alexsamardzic committed:
Update on "Fix dynamic shapes test failures on SM 7.x or lower."


For details, see: https://github.com/pytorch/pytorch/issues/122747.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

alexsamardzic committed:
Update on "Fix dynamic shapes test failures on SM 7.x or lower."


For details, see: https://github.com/pytorch/pytorch/issues/122747.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -135,6 +135,20 @@
+    def __call__(self, fn):
+        def wrapper(slf, *args, **kwargs):
+            fn._expected_failure_codegen_dynamic = (
+                self.test_instance.device == "cuda" and not SM80OrLater
```
Shouldn't this be `slf.device`. I would expect the decorator is only called once on the parent test class, and not for the device specific subclasses. Then again, somehow CI is passing so maybe not???

alexsamardzic replied:
You're right, that had the test passing as an accident.  I simplified it so that I hope it should work properly now, and it should be easier to understand.  But even this is not satisfactory, it depends on the sequential execution of tests.  Correct solution is probably just to duplicate the test code, decorate one copy with `onlyCpu` and the other one with `onlyCUDA` and `expectedFailureCodegenDynamic`, but only if `SM80OrLater`.

alexsamardzic committed:
Update on "Fix dynamic shapes test failures on SM 7.x or lower."


For details, see: https://github.com/pytorch/pytorch/issues/122747.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

alexsamardzic committed:
Update

[ghstack-poisoned]

tinglvv commented:
Hi @alexsamardzic , we are still seeing the dynamic shapes test failures on SM70 and lower in our internal CI, it is on mixed_mm3 now though. 
test signature
```
==================================== RERUNS ====================================
_______ DynamicShapesCodegenGPUTests.test_mixed_mm3_dynamic_shapes_cuda ________
Traceback (most recent call last):
  File "/usr/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/usr/local/lib/python3.10/dist-packages/torch/testing/_internal/common_utils.py", line 2979, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/pytorch/test/inductor/test_torchinductor.py", line 11384, in new_test
    return value(self)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/testing.py", line 320, in _fn
    return fn(*args, **kwargs)
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/opt/pytorch/pytorch/test/inductor/test_torchinductor.py", line 3290, in test_mixed_mm3
    self.common(
  File "/opt/pytorch/pytorch/test/inductor/test_torchinductor_codegen_dynamic_shapes.py", line 422, in common
    return check_codegen(
  File "/opt/pytorch/pytorch/test/inductor/test_torchinductor_codegen_dynamic_shapes.py", line 82, in check_codegen
    self.assertTrue("def triton" in code, f"Failed to find triton kernel\n{code}")
  File "/usr/lib/python3.10/unittest/case.py", line 687, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true : Failed to find triton kernel
```

Could we revisit this PR to resolve it? 
I guess we can also skip the mixed_mm tests on SM70 and lower.

alexsamardzic commented:
> Hi @alexsamardzic , we are still seeing the dynamic shapes test failures on SM70 and lower in our internal CI, it is on mixed_mm3 now though.

This test is introduced in #127663, so I guess @AlnisM is in better position to answer your question.

An unrelated issue is that, after all the changes in the meantime, these tests now do not seem to exercise CUTLASS "back-end" for auto-tuning...

AlnisM commented:
The test expects triton code to be generated, but on SM70 and lower, the generation of triton code is skipped (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm.py#L693). So I think this test should be skipped on SM70 and lower.

alexsamardzic commented:
> The test expects triton code to be generated, but on SM70 and lower, the generation of triton code is skipped (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm.py#L693). So I think this test should be skipped on SM70 and lower.

Agreed, and I'd say all three of these tests should be disabled if `not SM80OrLater` - as these are mostly your changes, do you want to do a PR?

AlnisM commented:
> > The test expects triton code to be generated, but on SM70 and lower, the generation of triton code is skipped (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm.py#L693). So I think this test should be skipped on SM70 and lower.
> 
> Agreed, and I'd say all three of these tests should be disabled if `not SM80OrLater` - as these are mostly your changes, do you want to do a PR?

I created a PR https://github.com/pytorch/pytorch/pull/138183. But, I don't have access to an SM70 GPU, so I wasn't able to test these changes locally, and I also don't have permissions to run tests in CI.
